File: scrutiny/sdk/docs/source/_static/code-examples/eol_config1.py
from scrutiny.sdk.client import ScrutinyClient

hostname = 'localhost'
port = 1234
client = ScrutinyClient()
with client.connect(hostname, port, wait_status=True):    # Establish a websocket connection and wait for a first server status update
    client.wait_device_ready(timeout=5)
    
    # The following watch uses "aliases". For HIL testing, it is a good practice to keep 
    # the interface to the firmware abstracted to ensure the validity of this script across version of firmwares.
    # Aliases are defined in the SFD, the byproduct of the SCrutiny post-build toolchain.

    eeprom_config_size  = client.watch('/alias/eeprom_configurator/size')   # Maps to /var/global/eeprom_configurator/m_size
    eeprom_config_addr  = client.watch('/alias/eeprom_configurator/addr')   # Maps to /var/global/eeprom_configurator/m_addr
    eeprom_config_cmd   = client.watch('/alias/eeprom_configurator/cmd')    # Maps to /var/global/eeprom_configurator/m_cmd
    
    eeprom_config_return_code = client.watch('/alias/eeprom_configurator/return_code')      # Maps to /var/global/eeprom_configurator/m_last_return_code
  
    assembly_header_model       = client.watch('/alias/eeprom_configurator/assembly_header/model')      # Maps to /var/global/eeprom_configurator/m_assembly_header/model
    assembly_header_version     = client.watch('/alias/eeprom_configurator/assembly_header/version')    # Maps to /var/global/eeprom_configurator/m_assembly_header/version
    assembly_header_revision    = client.watch('/alias/eeprom_configurator/assembly_header/revision')   # Maps to /var/global/eeprom_configurator/m_assembly_header/revision
    assembly_header_serial      = client.watch('/alias/eeprom_configurator/assembly_header/serial')     # Maps to /var/global/eeprom_configurator/m_assembly_header/serial

    client.wait_new_value_for_all()
    
    print("Writing the assembly information into the device EEPROM")

    with client.batch_write():
        # Order of write is respected. Block until completion at exit of "with" block
        assembly_header_model.value = 1
        assembly_header_version.value = 2
        assembly_header_revision.value = 3
        assembly_header_serial.value = 0x12345678
        eeprom_config_addr.value = 0
        eeprom_config_cmd.value = 3     # WriteAssemblyHeader  
        # eeprom_config_cmd.value = 'WriteAssemblyHeader'   # With enum support

    eeprom_config_cmd.wait_value(0, timeout=5)  # The device will set this back to None once finished      
    # eeprom_config_cmd.wait_value('None', timeout=2)   # With enum support

    # Request the server for an update. Make sure we do not read an old value
    # We have no control on the order in which the values a polled by the server.
    eeprom_config_return_code.force_refresh() 

    assert eeprom_config_return_code.value_int == 0, "Write operation failed"

    print("Successfully written")

File: scrutiny/sdk/docs/source/_static/code-examples/eol_config1.cpp
// ----------------
// EEPROMDriver.hpp
#include <cstdint>

// Fictive driver, no need to know how it works.
class EEPROMDriver{
public:
    int read(uint8_t * const buf, uint32_t const addr, uint32_t const size) const;
    int write(uint8_t const * const buf, uint32_t const addr, uint32_t const size) const;
    int erase();
};

// ----------------
// EEPROmConfigurator.hpp

class EEPROMConfigurator{
public:
    static constexpr uint32_t BUFFER_SIZE=1024;
    enum class Command{
        None,
        Read,
        Write,
        WriteAssemblyHeader,
        Erase
    };

    struct AssemblyHeader {
        uint8_t model;
        uint8_t version;
        uint8_t revision;
        uint32_t serial;
    };

    EEPROMConfigurator(EEPROMDriver* const driver) : 
        m_buffer{},
        m_size{0},
        m_addr{0},
        m_buffer_size{BUFFER_SIZE},
        m_driver{driver},
        m_cmd{Command::None},
        m_assembly_header{},
        m_last_return_code{0}
        {}
    void process();
private:
    uint8_t m_buffer[BUFFER_SIZE];
    uint32_t m_size;
    uint32_t m_addr;
    uint32_t m_buffer_size;
    EEPROMDriver* const m_driver;
    Command m_cmd;
    AssemblyHeader m_assembly_header;
    int m_last_return_code;
};


// ----------------
// EEPROmConfigurator.cpp
#include <cstring>

void EEPROMConfigurator::process()
{
    // In this function, we do some action on the EEPROM based on m_cmd, a variable that will be controlled by Scrutiny
    if (m_cmd == Command::None){
        return;
    }
    m_last_return_code = 1;
    if (m_cmd == Command::WriteAssemblyHeader)
    {
        m_last_return_code = m_driver->write(reinterpret_cast<uint8_t*>(&m_assembly_header), m_addr, sizeof(m_assembly_header));
    } 
    else if (m_cmd == Command::Write)
    {
        if (m_size < BUFFER_SIZE)
        {
            m_last_return_code = m_driver->write(m_buffer, m_addr, m_size);
        }
    }
    else if (m_cmd == Command::Read)
    {
        if (m_size < BUFFER_SIZE)
        {
            m_last_return_code = m_driver->read(m_buffer, m_addr, m_size);
        }
    }
    else if (m_cmd == Command::Erase)
    {
        m_last_return_code = m_driver->erase();
    }
    m_cmd = Command::None;
}


//-----------------
// main.cpp

void init();
void run_application();
void update_scrutiny();

EEPROMDriver eeprom_driver;
#ifdef ENABLE_EOL_CONFIGURATOR
EEPROMConfigurator eeprom_configurator(&eeprom_driver);
#endif

int main()
{
    init();
    while(true)
    {
#ifdef ENABLE_EOL_CONFIGURATOR
        eeprom_configurator.process();
#endif
        run_application();
        update_scrutiny();
    }
}
